electricitymap/contrib/libs/models/events.py

```diff
@@ -0,0 +1,427 @@
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta, timezone
+from logging import Logger
+from typing import Any, Dict, List, Optional
+
+from pydantic import BaseModel, validator
+
+from electricitymap.contrib.config import EXCHANGES_CONFIG, ZONES_CONFIG, ZoneKey
+from electricitymap.contrib.libs.models.constants import VALID_CURRENCIES
+
+LOWER_DATETIME_BOUND = datetime(2000, 1, 1, tzinfo=timezone.utc)
+
+
+class ProductionMix(BaseModel):
+    """
+    Containts the production mix for a zone at a given time.
+    All values are in MW.
+    """
+
+    biomass: Optional[float] = None
+    coal: Optional[float] = None
+    gas: Optional[float] = None
+    geothermal: Optional[float] = None
+    hydro: Optional[float] = None
+    nuclear: Optional[float] = None
+    oil: Optional[float] = None
+    solar: Optional[float] = None
+    unknown: Optional[float] = None
+    wind: Optional[float] = None
+
+    def set_value(self, mode: str, value: float) -> None:
+        """
+        Sets the value of a production mode.
+        This can be used if the Production has been initialized empty
+        and is being filled in a loop.
+        """
+        self.__setattr__(mode, value)
+
+
+class StorageMix(BaseModel):
+    """
+    Containts the storage mix for a zone at a given time.
+    All values are in MW.
+    """
+
+    battery: Optional[float] = None
+    hydro: Optional[float] = None
+
+
+class Event(BaseModel, ABC):
+    """
+    An abstract class representing all types of electricity events that can occur in a zone.
+    forecasted: Whether the point is a forecasted point or not.
+    Should be set to True if the point is a forecast provided by a datasource.
+    zoneKey: The zone key of the zone the event is happening in.
+    datetime: The datetime of the event.
+    source: The source of the event.
+    We currently use the root url of the datasource. Ex: edf.fr
+    """
+
+    forecasted: bool = False
```

Comment:
or simply a type field that can take values from an enum "measured", "forecasted", "estimated"?

Reply:
We have ruled out making children classes for this because it seemed a bit overkilled, since there shouldn't be much difference between the two. Though it might come in handy since we have plans to enhance the source.
An enum could be a nice solution though are we sure that "measured", "forecasted", "estimated" will always be mutually exclusive ?
Could it be that at some point we make an estimate of a forecasted point from an external provider ?

Reply:
Let's go for an enum.